sstables/trie/bti_key_translation.hh

```diff
@@ -0,0 +1,211 @@
+/*
+ * Copyright (C) 2024-present ScyllaDB
+ */
+
+/*
+ * SPDX-License-Identifier: LicenseRef-ScyllaDB-Source-Available-1.0
+ */
+
+// This file provides translation routines for ring positions and clustering positions
+// from Scylla's native in-memory structures to BTI's byte-comparable encoding.
+//
+// This translation is performed whenever a new decorated key or clustering block
+// are added to a BTI index, and whenever a BTI index is queried for a range of positions.
+//
+// For a description of the encoding, see
+// https://github.com/apache/cassandra/blob/fad1f7457032544ab6a7b40c5d38ecb8b25899bb/src/java/org/apache/cassandra/utils/bytecomparable/ByteComparable.md#multi-component-sequences-partition-or-clustering-keys-tuples-bounds-and-nulls
+
+#pragma once
+
+#include "mutation/position_in_partition.hh"
+#include "types/comparable_bytes.hh"
+#include "common.hh"
+#include "dht/ring_position.hh"
+#include <generator>
+#include "comparable_bytes_iterator.hh"
+
+namespace sstables {
+    struct clustering_info;
+} // namespace sstables
+
+namespace sstables::trie {
+
+// FIXME: this doesn't deserve to be a template.
+// The actual data structure is the same for both partition keys and clustering keys.
+// There's no need to codegen two separate functions for them.
+// Unfortunately p.components(representation).begin() returns an internal type
+// of compound_type. So this isn't easy to detemplatize.
```

Comment:
We could/should refactor compound_type to have a private non-template base class with most of the implementation, and add an accessor to gain access to that private base class for user that know what they are doing. Anyway not for now.